### ISSUE TITLE
[AIR] Composite Trainer

### ISSUE THREAD

**Issue by mlguruz** [ray-project/ray#28516]
### Description

So far, it seems that a Trainer is standalone. 
It would be nice if a Trainer can take other trainers as inputs, or, a DAG of trainers as inputs, in which case it acts as a "Composite Trainer" where `training_loop` is basically just executing the workflow based on the DAG.

### Use case

This will allow tuning multiple models (or basically any sub-graph) together possible.

**Comment by richardliaw** [comment]
Interesting... could you show me a use case?

**Comment by mlguruz** [comment]
> Interesting... could you show me a use case?

An example could be that you wanna fit three models that depend on the same preprocessor (e.g. some winsorizer) and a stacker that depends on the predictions from the three models.

In this case, 
1) the preprocessor should be fitted once 
2) there's no dep among the three models, so they could train in parallel, but they all depend on the preprocessor
3) the stacker depends on all three models so he needs to wait

So this is a DAG of models. 

I may wanna tune the preprocessor params allow w/ one of the model params (e.g. # of MADs + activation function).

**Comment by richardliaw** [comment]
So just so I understand this, is this basically:

```
dataset = ray.data.read_parquet()
preprocessor = Scaler()

preprocessor.fit(dataset)

result1 = SomeTrainer(preprocessor=preprocessor).fit(dataset)
result2 = SomeTrainer2(preprocessor=preprocessor).fit(dataset)
result3 = SomeTrainer3(preprocessor=preprocessor).fit(dataset)

predictor = create_stacked_predictor(result1.checkpoint, result2.checkpoint, result3.checkpoint)
predictor.predict()
```

> I may wanna tune the preprocessor params allow w/ one of the model params

What do you mean by this?

**Comment by mlguruz** [comment]
> So just so I understand this, is this basically:
> 
> ```
> dataset = ray.data.read_parquet()
> preprocessor = Scaler()
> 
> preprocessor.fit(dataset)
> 
> result1 = SomeTrainer(preprocessor=preprocessor).fit(dataset)
> result2 = SomeTrainer2(preprocessor=preprocessor).fit(dataset)
> result3 = SomeTrainer3(preprocessor=preprocessor).fit(dataset)
> 
> predictor = create_stacked_predictor(result1.checkpoint, result2.checkpoint, result3.checkpoint)
> predictor.predict()
> ```
> 
> > I may wanna tune the preprocessor params allow w/ one of the model params
> 
> What do you mean by this?
Yes on the pseudo-code. 

And sorry for the typo - I meant "along with"... 

So, I may wanna tune "max median-absolute-dev" values and "activation function" types together. They belong to two "models".

**Comment by richardliaw** [comment]
OK just help me understand -- when you're tuning the preprocessor with the model params, do all models get affected? What do you mean by they belong to "two models"

**Comment by mlguruz** [comment]
> OK just help me understand -- when you're tuning the preprocessor with the model params, do all models get affected? What do you mean by they belong to "two models"?


Yes - all downstream models will be affected. 

For (an extreme) example, when you set stdev of StandardScaler to 0, all feature values get clipped to mean, hence everything from that point on will be affected. 

I think a "model" as anything that takes in some data and spits out some data, so a preprocessor is the same as a NN (transform = predict, not identical, but you get my point).

**Comment by richardliaw** [comment]
OK got it, so you just want to be able to tune everything end-to-end. Kinda like the StackedRegressor in Scikit-Learn.

**Comment by mlguruz** [comment]
> OK got it, so you just want to be able to tune everything end-to-end. Kinda like the StackedRegressor in Scikit-Learn.

Yes - it avoids duplicated fits (like refitting the same preprocessors) and makes the whole dag tuneable. 

I have done that before but not on Ray, and now I wanna move it to Ray, since I really like the distributed training (w/ different resources for different models) features of Ray, and the Ray Tune ecosystem.  

Do you have any suggestions on how I could achieve that?

**Comment by richardliaw** [comment]
Hey sorry for the slow reply...

That's great to hear! To be honest it might be a bit hard at the moment. The reason is because we don't have great support for composing trainers together right now... @krfricke is working on something that will enable this type of workload.

Perhaps you can give us some pseudocode for how you might expect to express your tuning/stacking workload so that Kai will have a good idea + incorporate it in his new design? 

If you can do so, we can probably get back to you with a prototype in 2 weeks.

**Comment by mlguruz** [comment]
> Hey sorry for the slow reply...
> 
> That's great to hear! To be honest it might be a bit hard at the moment. The reason is because we don't have great support for composing trainers together right now... @krfricke is working on something that will enable this type of workload.
> 
> Perhaps you can give us some pseudocode for how you might expect to express your tuning/stacking workload so that Kai will have a good idea + incorporate it in his new design?
> 
> If you can do so, we can probably get back to you with a prototype in 2 weeks.

Not at all! Thank you for your help, and your timing couldn't be any better, as I just got a working example :)

Pls let me know if something like the following looks right to you: 


```
class DummyTrainer(BaseTrainer):
    _scaling_config_allowed_keys = BaseTrainer._scaling_config_allowed_keys + [
        "num_workers",
    ]
    
    def setup(self):
        self.scheduler = DaskOnRayScheduler(sync_mode=False, num_submit_threads=2)
        self.t1 = TorchTrainer(
            train_func,
            train_loop_config={
                "lr": 1e-3,
                "batch_size": 64,
                "epochs": 4
            },
            scaling_config=ScalingConfig(num_workers=2, use_gpu=False),
            run_config=RunConfig(name='test_run3'),
        )
        self.t2 = TorchTrainer(
            train_func,
            train_loop_config={
                "lr": 1e-3,
                "batch_size": 64,
                "epochs": 4
            },
            scaling_config=ScalingConfig(num_workers=2, use_gpu=False),
            run_config=RunConfig(name='test_run4'),
        )
    
    def training_loop(self) -> None:
        print("running training loop...")
        m1 = self.scheduler.add_task(self.t1.fit, fn_key='f1',)
        m2 = self.scheduler.add_task(self.t2.fit, fn_key='f2',)
        self.scheduler.compute([m1, m2])
```

Several notes:
1. the `TorchTrainer` is copying + pasting your example: https://docs.ray.io/en/latest/train/examples/torch_fashion_mnist_example.html
2. the `DaskOnRayScheduler` is something a bit proprietary but in essence its wrapping the `ray_dask_get` function in file ray/util/scheduler.py...

**Comment by richardliaw** [comment]
Honestly that's amazing that you got it to work... I would have thought there'd be some issues. A couple thoughts:

1. Maybe you can get away with simple threading rather than needing to invoke dask on ray.
2. How big is your machine/cluster?

Let's chat offline.
